returnn/tf/layers/basic.py

```diff
@@ -913,37 +913,204 @@ def transform_config_dict(cls, d, network, get_layer):
 
 class GatherLayer(_ConcatInputLayer):
   """
-  This is basically ``x[position]`` in some axis.
-  See also :class:`SliceLayer`.
-  See also :class:`GatherNdLayer`, which works on the time-axis, but supports different positions per batch.
+  Gathers slices on a specified axis from the input layer using indices from a ``position`` layer.
+  If the input is a layer of the shape ``[B,D,F1]``, and indices of shape ``[B,F2]``, this will yield output of the
+  shape ``[B,F2,F1]`` where
 
-  (Currently `position` can just be an integer, so it's much less generic than ``tf.gather``.
-   However, we might extend this to lists, arrays, or even other layers.)
+  ``output[b,f2,f1] = input[b,indices[b,f2],f1]``
+
+  (if ``D`` is the axis to gather from).
+  In general, all shared axes of the input and the positions will be considered as batch-axes.
+
+  It's basically a wrapper around ``tf.gather``.
+  See also :class:`GatherNdLayer`.
   """
   layer_class = "gather"
 
-  def __init__(self, axis, position, **kwargs):
+  def __init__(self, position, axis, **kwargs):
     """
-    :param str axis:
-    :param int position:
+    :param LayerBase|int position: Layer containing the indices used to select the slices of the input from.
+    If another layer, must be of type ``int32`` or ``int64``.
+    Can also specify a constant ``int``.
+    :param str axis: The axis into which we gather the indices into
     """
     super(GatherLayer, self).__init__(**kwargs)
-    axis = self.input_data.get_axis_from_description(axis, allow_int=False)
-    assert isinstance(position, int)  # nothing else implemented yet
-    slices = [slice(None, None)] * axis + [position]  # type: typing.List[typing.Union[slice, int]]
-    self.output.placeholder = self.input_data.placeholder[slices]
+    self.position = position
+
+    input_data = self.input_data
+    if isinstance(position, int):
+      constant_kwargs = dict(name="constant_position", network=kwargs["network"], value=position, dtype='int32')
+      position = ConstantLayer(output=ConstantLayer.get_out_data_from_opts(**constant_kwargs), **constant_kwargs)
```

Comment:
Do not create a dummy layer. This is ugly. And no real need for it.

You could operate on `Data` directly. Use `Data.from_tensor`.